### PR TITLE
Kinesis Connector Integration

### PR DESCRIPTION
**Pull request by shubham166**
Issue :  Kinesis connector #2366 
Presto Connector to Amazon Kinesis.

### CODE REVIEW

**Comment by electrum** [comment]
Thanks for sending this! There seems to be a lot of duplication with the Kafka connector, especially around the decoding logic for rows/fields. How much of this could we pull out into a shared module? I think this could be useful for other connectors, too.

**Comment by shubham166** [comment]
Yes, decoder logic for Kafka and Kinesis is almost same. It would be a really great idea to pull the decoder module to shared module, which can be used for other connectors in future.

**Comment by snarayananqubole** [comment]
There is also an open PR for redis connector. Might be worth doing refactoring as a separate PR?

**Comment by losipiuk** [comment]
looking into this one

**Comment by snarayanank2** [comment]
The latest version of the code is available [here](https://github.com/qubole/presto-kinesis). Feel free to fork / send patches there.

Here's an [AWS blogpost] (https://blogs.aws.amazon.com/bigdata/post/Tx2DDFNHXSAAH2G/Presto-Amazon-Kinesis-Connector-for-Interactively-Querying-Streaming-Data).

File: presto-docs/src/main/sphinx/connector/kinesis.rst [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+..note::
+
+    For textual topics that contain JSON data, it is entirely possible to not
+    use any table definition files, but instead use the Presto
+    :doc:`/functions/json` to parse the ``_message`` column which contains
+    the bytes mapped into an UTF-8 string. This is, however, pretty
+    cumbersome and makes it difficult to write SQL queries.
+
+A table definition file consists of a JSON definition for a table. The
+name of the file can be arbitrary but must end in ``.json``.
```
Comment (losipiuk) [inline presto-docs/src/main/sphinx/connector/kinesis.rst]: must end with ``.json``.

File: presto-docs/src/main/sphinx/connector/kinesis.rst [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+
+.. code-block:: json
+
+    {
+        "tableName": ...,
+        "schemaName": ...,
+        "streamName": ...,
+        "message": {
+            "dataFormat": ...,
+            "field": [
```
Comment (losipiuk) [inline presto-docs/src/main/sphinx/connector/kinesis.rst]: Further part of doc suggests that this should be `fields`.

File: presto-kinesis/pom.xml [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>com.facebook.presto</groupId>
+        <artifactId>presto-root</artifactId>
+        <version>0.107-SNAPSHOT</version>
+    </parent>
+
+
```
Comment (losipiuk) [inline presto-kinesis/pom.xml]: Just remove commented sections from this file.

File: presto-docs/src/main/sphinx/connector/kinesis-tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,252 @@
+===========================
```
Comment (losipiuk) [inline presto-docs/src/main/sphinx/connector/kinesis-tutorial.rst]: sphinx compiler fails when generating docs. Please fix that.

File: presto-kinesis/pom.xml [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+      <scope>test</scope>
+    </dependency>  -->
+
+    <dependency>
+        <groupId>org.weakref</groupId>
+        <artifactId>jmxutils</artifactId>
+        <scope>provided</scope>
+    </dependency>
+
+    <dependency>
```
Comment (losipiuk) [inline presto-kinesis/pom.xml]: do not use `provided` scope for libraries which are not used as part of presto<->presto-connector interface.

File: presto-kinesis/pom.xml [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+            <groupId>io.airlift</groupId>
```
Comment (losipiuk) [inline presto-kinesis/pom.xml]: why those excludes.
Why not simple dependency definition as:
```
        <dependency>
            <groupId>io.airlift</groupId>
            <artifactId>bootstrap</artifactId>
        </dependency>
```

File: presto-kinesis/pom.xml [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+                </exclusion>
+                <exclusion>
+                    <groupId>org.slf4j</groupId>
+                    <artifactId>jcl-over-slf4j</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+
+        <dependency>
+            <groupId>io.airlift</groupId>
```
Comment (losipiuk) [inline presto-kinesis/pom.xml]: ditto

**Comment by losipiuk** [comment]
General note:
There is a lot of codestyle violation errors. Just run `mvn install`.
If you are using IntelliJ there is codestyle config available here: https://github.com/airlift/codestyle

**Comment by losipiuk** [comment]
Please rebase to master if possible - there were some changes in the spi.

File: presto-kinesis/src/main/java/com/facebook/presto/kinesis/KinesisMetadata.java [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+
+public class KinesisMetadata
+        extends ReadOnlyConnectorMetadata
```
Comment (losipiuk) [inline presto-kinesis/src/main/java/com/facebook/presto/kinesis/KinesisMetadata.java]: ReadOnlyConnectorMetdata is now gone and replaced with default methods in ConnectorMetadata interface.

File: presto-kinesis/src/main/java/com/facebook/presto/kinesis/KinesisAwsCredentials.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+ */
+package com.facebook.presto.kinesis;
+
+import com.amazonaws.auth.AWSCredentials;
+import com.google.inject.Inject;
+
+public class KinesisAwsCredentials
+        implements AWSCredentials
+{
+    private String accessKeyId;
```
Comment (losipiuk) [inline presto-kinesis/src/main/java/com/facebook/presto/kinesis/KinesisAwsCredentials.java]: make those final

File: presto-kinesis/src/main/java/com/facebook/presto/kinesis/KinesisAwsCredentials.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    private String secretKey;
+
+    @Inject
+    public KinesisAwsCredentials(String accessKeyId, String secretKey)
```
Comment (losipiuk) [inline presto-kinesis/src/main/java/com/facebook/presto/kinesis/KinesisAwsCredentials.java]: Remove `@Inject`. If you wanted to inject accessKey and secretKey you would need to annotate parameters somehow. Binding by type does not work as both are Strings. And you do not use dependency injection mechanism for this class anyway.

File: presto-kinesis/pom.xml [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+            <groupId>io.airlift.tpch</groupId>
+            <artifactId>tpch</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+  </dependencies>
+
+  <build>
+        <plugins>
+            <plugin>
```
Comment (losipiuk) [inline presto-kinesis/pom.xml]: Not sure why you need this plugin definition. It is defined in root already.

**Comment by martint** [comment]
> The latest version of the code is available here. Feel free to fork / send patches there

@snarayanank2, this code was originally contributed by someone in your team, right? Are you still interested in getting it in given that you have a separate project to develop it? If not, we can just close this pull request.

**Comment by snarayanank2** [comment]
@martint - I'm not affiliated with Qubole anymore. @stagraqubole would be the right person to answer your question.

**Comment by stagraqubole** [comment]
@martint we would like to have this in trunk. This pull request now needs quite a few changes to utilize the new presto-record-decoder module and to take care of comments by @losipiuk.

The original author has also moved on and we cannot use this to pull request for further changes. We can close this one and in sometime we will open a new request that handles the suggested changes.

**Comment by martint** [comment]
@stagraqubole Sounds good. Let's do that, then.

**Pull request closed without merging**